client/src/components/general/VirtualizedList.tsx

```diff
@@ -0,0 +1,80 @@
+import { useCallback, useEffect, useRef, useState } from "react";
+import styles from './VirtualizedList.module.scss';
+import { TransitionGroup, CSSTransition } from 'react-transition-group';
+import { ListItem } from "../../models/ListItem";
+
+type Props = {
+    items: ListItem[],
+    itemHeight: number
+}
+
+export function VirtualizedList(props: Props) {  
+  const { items, itemHeight } = props;
+  const [selectedItem, setSelectedItem] = useState<number | null>(null);
+  const [visibleItems, setVisibleItems] = useState<ListItem[]>([]);
+  const [scrollTop, setScrollTop] = useState(0);
+  
+  const container = useRef<HTMLDivElement>(null);
+  
+  const changeBackgroundColor = useCallback((id: number) => {
+    setSelectedItem(id);
+  }, []);
+
+  const updateScroll = () => {
+    if (container.current) {
+      setScrollTop(container.current.scrollTop);
+    }
+  };
+    
+  useEffect(() => {
+    if (container.current) {
+      container.current.addEventListener('scroll', updateScroll);
+      return () => container.current?.removeEventListener('scroll', updateScroll);
+    }
+  }, []);
+
+  useEffect(() => {
+    // Calculating items to render
+    const containerHeight = container.current?.clientHeight || 0;
+    const startIndex = Math.floor(scrollTop / itemHeight);
+    const endIndex = Math.min(
```

Comment:
this can be calculated outside the useEffect at render time

Reply:
changed it and calculated startIndex and containerHeight at render time, full explanation in above :)